filament/src/fg/ResourceAllocator.cpp

```diff
@@ -40,14 +76,92 @@ void ResourceAllocator::destroyRenderTarget(RenderTargetHandle h) noexcept {
     return mBackend.destroyRenderTarget(h);
 }
 
-TextureHandle ResourceAllocator::createTexture(SamplerType target, uint8_t levels,
-        TextureFormat format, uint8_t samples, uint32_t width, uint32_t height,
-        uint32_t depth, TextureUsage usage) noexcept {
-    return mBackend.createTexture(target, levels, format, samples, width, height, depth, usage);
+backend::TextureHandle ResourceAllocator::createTexture(const char* name,
+        backend::SamplerType target, uint8_t levels,
+        backend::TextureFormat format, uint8_t samples, uint32_t width, uint32_t height,
+        uint32_t depth, backend::TextureUsage usage) noexcept {
+    // do we have a suitable texture in the cache?
+    TextureHandle handle;
+    if (mEnabled) {
+        auto& textureCache = mTextureCache;
+        const TextureKey key{ name, target, levels, format, samples, width, height, depth, usage };
+        auto it = textureCache.find(key);
+        if (UTILS_LIKELY(it != textureCache.end())) {
+            // we do, move the entry to the in-use list, and remove from the cache
+            handle = it->second.handle;
+            mCacheSize -= it->second.size;
+            textureCache.erase(it);
+        } else {
+            // we don't, allocate a new texture and populate the in-use list
+            handle = mBackend.createTexture(
+                    target, levels, format, samples, width, height, depth, usage);
+        }
+        mInUseTextures.emplace(std::make_pair(handle, key));
+    } else {
+        handle = mBackend.createTexture(
+                target, levels, format, samples, width, height, depth, usage);
+    }
+    return handle;
 }
 
 void ResourceAllocator::destroyTexture(TextureHandle h) noexcept {
-    return mBackend.destroyTexture(h);
+    if (mEnabled) {
+        // find the texture in the in-use list (it must be there!)
+        auto it = mInUseTextures.find(h);
+        assert(it != mInUseTextures.end());
+
+        // move it to the cache
+        const TextureKey key = it->second;
+        uint32_t size = key.getSize();
+
+        mTextureCache.insert(std::make_pair(key, TextureCachePayload{ h, mAge, size }));
```

Comment:
In modern C++ this can simply be `mTextureCache.insert({key, TextureCachePayload{ h, mAge, size }})`